### DISCUSSION TITLE
Are the "LEGACY" cookies still usefull ?

### DISCUSSION THREAD

**Discussion by fidgi** [keycloak/keycloak#8938] [Q&A] [answered]
Hi,

What is the purpose of the cookies with the "_LEGACY" suffix that are just a duplicate of the cookies w/ the same name but w/o the suffix ?

Is there a way to disable the generation of these cookies?

FYI, We're using Keycloak 14.0.0 .

Regards,

**Comment by vmuzikar** [comment, marked as answer]
Hello,
yes, there are still use cases for it. They have different attributes.

1. Non-legacy cookies have SameSite=None set; legacy cookies don't have SameSite set at all.
2. Non-legacy cookies always have Secure flag, i.e. can be accessed only via SSL (this is mandatory for SameSite=None); legacy cookies respect the Realm setting for the Secure attribute (that is by default only for external requests is SSL required to access the cookies)

Non-legacy cookies are used in 3rd party contexts, when KC cookies need to be accessed from another hostname, typically by the JS adapter.

Legacy cookies are required as a workaround for older Apple devices that had a bug [1] in SameSite handling. Also, thanks to the variable Secure attribute the cookies are useful when running KC locally during development as SSL is not required in this case. However, they are inaccessible in 3rd party contexts. Thanks to this they don't possess any security risk, together with the Secure attribute when accessed externally, and HttpOnly attribute for any sensitive cookies.

[1] E.g. https://developer.apple.com/forums/thread/658688

**Reply by stianst** [reply]
@vmuzikar wonder if we really should rather re-design this stuff, and detect specific browser versions to just set the single cookie correctly rather than having the complexity of the two duplicated cookies.

**Reply by vmuzikar** [reply]
As I mentioned, it has advantages for all browsers when you don't want to use SSL (e.g. for dev purposes).

**Reply by stianst** [reply]
@vmuzikar That's something we can detect as well, and configure SameSite accordingly. I essentially think that we should have a single cookie and be smarter on how we set it on the server side, rather than having the two cookies.

**Reply by vmuzikar** [reply]
IMHO there's no harm in having two cookies but I agree a single one is a cleaner solution as we tend to be asked about it quite a lot.

I would not try to detect the browser, though. It's quite a complex task add additional complexity (or new dependencies) just for a workaround for old Apple OS's. Apple updates their devices quite well and the affected OS's/browsers have even lower market share since when we implemented the workaround.

I can imagine we set the `SameSite` attribute and therefore also `Secure` attribute based on the HTTP/HTTPS request. Maybe we could even somehow leverage the Quarkus profiles, maybe have a little bit less strict behavior in the dev profile?

**Reply by stianst** [reply]
What browsers do need a workaround today? We already have dependencies on parsing the user-agent headers, so couldn't that be used relatively easily?

For the Secure that's pretty much just setting it correctly if it's a http or https request. Whether or not http is permitted or not is controlled in Quarkus by running prod or dev profiles, or explicitly enabling http, which further can be controlled by the proxy settings if it's actually https to the user, or just to the proxy. The bit setting the cookie should just set the value correctly depending if it's a http or https request though, as the rest should be enforced elsewhere.

Oh, and the sslrequired option in a realm should probably go away ;)

**Reply by vmuzikar** [reply]
The workaround is needed only for [older Apple browsers/OSs](https://caniuse.com/?search=samesite) (notice that even 3rd party browsers on iOS/iPadOS use the engine from Safari).

We don't have much choice around `Secure`. When using `SameSite=None`, we also [need to use `Secure`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite).
